File: LWFileDownload/Classes/LWFileDownloadManager.m
//
// Created by Luo Wei on 2017/11/25.
// Copyright (c) 2017 wodedata. All rights reserved.
//

#import "LWFileDownloadManager.h"

@interface LWFileDownloadManager ()

@property(nonatomic, copy) NSString *diretoryName;

@end

@implementation LWFileDownloadManager {
    NSString *_fileDirectoryPath;
}

static LWFileDownloadManager *_instance = nil;

+ (instancetype)shareManager {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _instance = [[self alloc] init];
    });
    return _instance;
}

+(int)alreadyExsitCode{
    return 300;
}

-(NSMutableDictionary *)taskMap{
    if(!_taskMap){
        _taskMap = @{}.mutableCopy;
    }
    return _taskMap;
}

//根据文件名与下载地址，获得路径
+ (NSString *)filePathWithFileName:(NSString *)fileName downloadURLString:(NSString *)urlString
              requestBlock:(NSMutableURLRequest *(^)(NSMutableURLRequest *,LWFileDownloadTask *))requestHandleBlock
             progressBlock:(void (^)(float,LWFileDownloadTask *))updateProgressBlock
             completeBlock:(void (^)(NSError *,LWFileDownloadTask *))serialCompleteBlock {

    NSString *filePath = [LWFileDownloadManager filePathWithFileName:fileName];

    BOOL exsit = [LWFileDownloadManager downloadFileWithFileName:fileName urlString:urlString requestBlock:requestHandleBlock progressBlock:updateProgressBlock completeBlock:serialCompleteBlock];

    if(!exsit){    //从本地bundle中获取
        filePath = [[[NSBundle bundleForClass:self.class] resourcePath] stringByAppendingPathComponent:fileName];
    }
    return filePath;

}

//判断是存已存在，如果不存在就下载,返回下载后的文件路径
+ (BOOL)downloadFileWithFileName:(NSString *)fileName urlString:(NSString *)urlString
                          requestBlock:(NSMutableURLRequest *(^)(NSMutableURLRequest *,LWFileDownloadTask *))requestHandleBlock
                         progressBlock:(void (^)(float,LWFileDownloadTask *))updateProgressBlock
                         completeBlock:(void (^)(NSError *,LWFileDownloadTask *))serialCompleteBlock{

    BOOL exsit = [LWFileDownloadManager exsitFileWithFileName:fileName];
    if (!exsit) {

        //更新request
        NSMutableURLRequest * (^updateRequest)(NSMutableURLRequest *,LWFileDownloadTask *)=^NSMutableURLRequest *(NSMutableURLRequest *request,LWFileDownloadTask *task){
            NSMutableURLRequest *req = request;
            if(requestHandleBlock){
                req = requestHandleBlock(request,task);
            }
            return req;
        };

        void (^progressBlock)(float,LWFileDownloadTask *) =^(float progress,LWFileDownloadTask *task) {
            if(updateProgressBlock){
                updateProgressBlock(progress,task);
            }
        };

        //完成
        void (^completeBlock)(NSError *,LWFileDownloadTask *)=^(NSError *error,LWFileDownloadTask *task) {
            if(serialCompleteBlock){
                serialCompleteBlock(error,task);
            }
        };

        //在这里把tsk存入字典队列
        LWFileDownloadTask *tsk = [LWFileDownloadManager
                downloadWithFileName:fileName
                           urlString:urlString
                        requestBlock:^NSMutableURLRequest *(NSMutableURLRequest *request) {
                            request = updateRequest(request, tsk);
                            return request;
                        }
                       progressBlock:^(float progress) {
                           progressBlock(progress, tsk);
                       }
                       completeBlock:^(NSError *error){
                           completeBlock(error,tsk);
                           [[LWFileDownloadManager shareManager].taskMap removeObjectForKey:urlString];  //从任务队列删除
                       }];

        //如果队列中没有任务就加入，并开始
        if(![LWFileDownloadManager shareManager].taskMap[urlString]){
            [LWFileDownloadManager shareManager].taskMap[urlString]=tsk;  //加入任务队列
            [tsk start];
        }

    }

    return exsit;
}

//并行下载,串行回调CompleteBlock //DISPATCH_QUEUE_CONCURRENT DISPATCH_QUEUE_SERIAL
+ (LWFileDownloadTask *)downloadWithFileName:(NSString *)fileName urlString:(NSString *)urlString
                                requestBlock:(NSMutableURLRequest *(^)(NSMutableURLRequest *))requestHandleBlock
                               progressBlock:(void (^)(float))updateProgressBlock
                               completeBlock:(void (^)(NSError *))serialCompleteBlock {

    //构造manager
    LWFileDownloadManager *manager = [LWFileDownloadManager shareManager];
    if (!manager.group) {
        manager.group = dispatch_group_create();
    }
    if (!manager.s_queue) {
        manager.s_queue = dispatch_get_main_queue();
//        self.s_queue = dispatch_queue_create("com.koou.com.group.once.queue", DISPATCH_QUEUE_SERIAL);
    }

    //更新request
    NSMutableURLRequest * (^updateRequest)(NSMutableURLRequest *)=^NSMutableURLRequest *(NSMutableURLRequest *request){
        NSMutableURLRequest *req = request;
        if(requestHandleBlock){
            req = requestHandleBlock(request);
        }
        return req;
    };

    //完成
    void (^completeBlock)(NSError *)=^(NSError *error){
        //把block放到group中顺序执行
        dispatch_group_async(manager.group, manager.s_queue, ^{
            if(serialCompleteBlock){
                serialCompleteBlock(error);
            }
        });
    };

    //新建任务
    LWFileDownloadTask *task = [LWFileDownloadTask taskWithURLString:urlString fileName:fileName];
    task.updateRequest = updateRequest;
    task.updateProgessBlock = updateProgressBlock;
    task.completeBlock = completeBlock;
//    [task start];  //开始下载

//    dispatch_group_notify(manager.group, manager.s_queue, ^{});
    return task;
}


#pragma mark - Helper Method

- (NSString *)fileDirectoryPath {

    NSString *documentsDirectory = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES)[0];
    _fileDirectoryPath = [documentsDirectory stringByAppendingPathComponent:_diretoryName ?: @"data"];

    [LWFileDownloadManager createDirectoryIfNotExsitPath:_fileDirectoryPath];   //创建目录

    LWDLLog(@"======fileDirectoryPath:%@",_fileDirectoryPath);
    LWDLLog(@"======App Bundle Path:%@",[[NSBundle mainBundle] bundlePath]);
    LWDLLog(@"======Home Path:%@",NSHomeDirectory());

    return _fileDirectoryPath;
}

//删除指定目录的文件
+ (BOOL)removeFileWithFilePath:(NSString *)filePath {
    //把文件删除
    NSError *err = nil;
    BOOL result = [[NSFileManager defaultManager] removeItemAtPath:filePath error:&err];
    if(err){
        LWDLLog(@"Error! %@", err);
    }
    return result;
}

//写入数据到指定路径
+ (BOOL)writeData:(NSData *)data toFilePath:(NSString *)filePath {
    BOOL result = [data writeToFile:filePath atomically:YES];  //写入到文件
    LWDLLog(@"=======file writeToFile:%@",(result ? @"YES" : @"NO"));
    return result;
}

//获取document下的指定path的绝对路径
+ (NSString *)documentDirectoryPath:(NSString *)path {
    NSString *documentsDirectory = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES)[0];
    NSString *directoryPath = [documentsDirectory stringByAppendingPathComponent:path];

    [LWFileDownloadManager createDirectoryIfNotExsitPath:directoryPath];
    return directoryPath;
}


//创建目录
+ (BOOL)createDirectoryIfNotExsitPath:(NSString *)path {
    BOOL success = YES;
    if(![[NSFileManager defaultManager] fileExistsAtPath:path]){  //如果则创建文件夹
        NSError * error = nil;
        success = [[NSFileManager defaultManager] createDirectoryAtPath:path withIntermediateDirectories:YES attributes:nil error:&error];
        if (!success || error) {
            LWDLLog(@"Error! %@", error);
        } else {
            LWDLLog(@"Create fonts directory Success!");
        }
    }
    return success;
}


//是否存在fileName
+ (BOOL)exsitFileWithFileName:(NSString *)fileName {
    NSString *filePath = [[LWFileDownloadManager shareManager].fileDirectoryPath stringByAppendingPathComponent:fileName];
    BOOL exsit = [[NSFileManager defaultManager] fileExistsAtPath:filePath];
    return exsit;
}

+(NSString *)filePathWithFileName:(NSString *)fileName {
    NSString *filePath = [[LWFileDownloadManager shareManager].fileDirectoryPath stringByAppendingPathComponent:fileName];
    return filePath;
}

@end





#pragma mark - LWFileDownloadTask Concurrence Download

@implementation LWFileDownloadTask


+ (instancetype)taskWithURLString:(NSString *)urlString fileName:(NSString *)fileName {
    LWFileDownloadTask *task = [[LWFileDownloadTask alloc] init];
    task.urlString = urlString;
    task.fileName = fileName;
    return task;
}

#pragma mark -

-(void)start{
    [self downloadCustomFileWithfileName:self.fileName URLString:self.urlString];
}


//下载
- (void)downloadCustomFileWithfileName:(NSString *)fileName URLString:(NSString *)urlString {

    self.fileName = fileName;
    BOOL exsit = [LWFileDownloadManager exsitFileWithFileName:fileName];
    if (exsit) {  //如果已经下载过了
        //更新UI
        if(self.updateProgessBlock){
            self.updateProgessBlock(1.0f);
        }
        if(self.completeBlock){
            NSError *error=[NSError errorWithDomain:@"already exsit or downloaded" code:[LWFileDownloadManager alreadyExsitCode] userInfo:@{@"filename": fileName, @"url": urlString, @"msg": @"已经下载过了"}];
            self.completeBlock(error);
        }
        return;
    }

    //构造NSURLSession
    //urlString = [urlString stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLQueryAllowedCharacterSet]];
    NSURL *url = [NSURL URLWithString:urlString];
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];

    [request setValue:@"http://app.wodedata.com" forHTTPHeaderField:@"Referer"];

    if(self.updateRequest){
        request = self.updateRequest(request);
    }

    NSURLSessionConfiguration *defaultConfigObject = [NSURLSessionConfiguration defaultSessionConfiguration];
    NSURLSession *defaultSession = [NSURLSession sessionWithConfiguration:defaultConfigObject
                                                                 delegate:self
                                                            delegateQueue:[NSOperationQueue mainQueue]];

    if(self.curretnDataTask && self.curretnDataTask.state != NSURLSessionTaskStateCompleted){   //取消原来的任务
        [self.curretnDataTask cancel];
    }
    self.curretnDataTask = [defaultSession dataTaskWithRequest:request];
    [self.curretnDataTask resume];

}

#pragma mark - NSURLSessionDataDelegate

- (void)URLSession:(NSURLSession *)session dataTask:(NSURLSessionDataTask *)dataTask didReceiveResponse:(NSURLResponse *)response
 completionHandler:(void (^)(NSURLSessionResponseDisposition disposition))completionHandler {
    LWDLLog(@"--------%d:%s \n", __LINE__, __func__);
    completionHandler(NSURLSessionResponseAllow);

    self.progress = 0.0f;
    self.downloadSize = [response expectedContentLength];
    self.dataToDownload = [[NSMutableData alloc] init];

//    dispatch_async(dispatch_get_main_queue(), ^{
//        if(self.showProgressBlock){ //显示下载进度提示,开始显示下载进度
//            self.showProgressBlock();
//        }
//    });
}

- (void)URLSession:(NSURLSession *)session dataTask:(NSURLSessionDataTask *)dataTask didReceiveData:(NSData *)data {
    LWDLLog(@"--------%d:%s \n", __LINE__, __func__);

    [self.dataToDownload appendData:data];
    self.progress = (float) [self.dataToDownload length] / self.downloadSize;
    LWDLLog(@"=======progress:%.4f, dataToDownload:%lli, downloadSize:%lli", self.progress,
            (long long int) self.dataToDownload.length, self.downloadSize);

    dispatch_async(dispatch_get_main_queue(), ^{
        if (self.updateProgessBlock) {
            self.updateProgessBlock(self.progress >= 1 ? 1 : self.progress);
        }

        if (self.progress < 1) {
            return;
        }

    });
}


- (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didCompleteWithError:(NSError *)error {
    LWDLLog(@"--------%d:%s \n", __LINE__, __func__);
    LWDLLog(@"=====completed; error: %@", error);

    BOOL exsit = [LWFileDownloadManager exsitFileWithFileName:self.fileName];
    if (!error && !exsit) {
        //写入文件
        NSString *filePath = [[LWFileDownloadManager shareManager].fileDirectoryPath stringByAppendingPathComponent:self.fileName];
        [LWFileDownloadManager writeData:self.dataToDownload toFilePath:filePath];
    }

    dispatch_async(dispatch_get_main_queue(), ^{
        //更新UI
        if (self.completeBlock) {
            self.completeBlock(error);
        }
    });

}

@end

File: LWFileDownload_swift/Classes/LWFileDownloadManager.swift
//
// Created by Luo Wei on 2017/11/25.
// Copyright (c) 2017 wodedata. All rights reserved.
// Swift version created 2025
//

import Foundation

#if DEBUG
func LWDLLog(_ format: String, _ args: CVarArg...) {
    NSLog(format, args)
}
#else
func LWDLLog(_ format: String, _ args: CVarArg...) {}
#endif

/// Main file download manager class - Singleton pattern
@objc public class LWFileDownloadManager: NSObject {

    // MARK: - Properties

    public let group: DispatchGroup
    public let serialQueue: DispatchQueue

    private var taskMap: [String: LWFileDownloadTask] = [:]
    private var directoryName: String = "data"
    private var _fileDirectoryPath: String?

    // MARK: - Singleton

    @objc public static let shared = LWFileDownloadManager()

    private override init() {
        self.group = DispatchGroup()
        self.serialQueue = DispatchQueue.main
        super.init()
    }

    @objc public static var alreadyExistCode: Int {
        return 300
    }

    // MARK: - Public API

    /// Get file path for a given fileName and download URL
    /// - Parameters:
    ///   - fileName: The name of the file
    ///   - urlString: The download URL string
    ///   - requestBlock: Optional block to modify the URLRequest
    ///   - progressBlock: Progress update callback
    ///   - completeBlock: Completion callback
    /// - Returns: The file path (either from documents directory or bundle)
    @objc public static func filePath(
        withFileName fileName: String,
        downloadURLString urlString: String,
        requestBlock: ((URLRequest, LWFileDownloadTask) -> URLRequest)? = nil,
        progressBlock: ((Float, LWFileDownloadTask) -> Void)? = nil,
        completeBlock: ((Error?, LWFileDownloadTask) -> Void)? = nil
    ) -> String {

        var filePath = LWFileDownloadManager.filePath(withFileName: fileName)
        let exists = LWFileDownloadManager.downloadFile(
            withFileName: fileName,
            urlString: urlString,
            requestBlock: requestBlock,
            progressBlock: progressBlock,
            completeBlock: completeBlock
        )

        if !exists {
            // Get from bundle if not in documents
            filePath = Bundle(for: self).resourcePath?.appending("/\(fileName)") ?? filePath
        }

        return filePath
    }

    /// Download file if it doesn't exist
    /// - Parameters:
    ///   - fileName: The name of the file
    ///   - urlString: The download URL string
    ///   - requestBlock: Optional block to modify the URLRequest
    ///   - progressBlock: Progress update callback
    ///   - completeBlock: Completion callback
    /// - Returns: true if file already exists, false if download was initiated
    @objc @discardableResult
    public static func downloadFile(
        withFileName fileName: String,
        urlString: String,
        requestBlock: ((URLRequest, LWFileDownloadTask) -> URLRequest)? = nil,
        progressBlock: ((Float, LWFileDownloadTask) -> Void)? = nil,
        completeBlock: ((Error?, LWFileDownloadTask) -> Void)? = nil
    ) -> Bool {

        let exists = LWFileDownloadManager.fileExists(withFileName: fileName)
        if exists {
            return true
        }

        // Create task with closures that capture the task itself
        var task: LWFileDownloadTask!

        let updateRequestBlock: ((URLRequest) -> URLRequest) = { request in
            if let requestBlock = requestBlock {
                return requestBlock(request, task)
            }
            return request
        }

        let updateProgressBlock: ((Float) -> Void) = { progress in
            progressBlock?(progress, task)
        }

        let completeBlockWrapper: ((Error?) -> Void) = { error in
            completeBlock?(error, task)
            LWFileDownloadManager.shared.taskMap.removeValue(forKey: urlString)
        }

        task = LWFileDownloadManager.download(
            withFileName: fileName,
            urlString: urlString,
            requestBlock: updateRequestBlock,
            progressBlock: updateProgressBlock,
            completeBlock: completeBlockWrapper
        )

        // Add to task queue if not already present
        if LWFileDownloadManager.shared.taskMap[urlString] == nil {
            LWFileDownloadManager.shared.taskMap[urlString] = task
            task.start()
        }

        return false
    }

    // MARK: - Private Methods

    private static func download(
        withFileName fileName: String,
        urlString: String,
        requestBlock: ((URLRequest) -> URLRequest)? = nil,
        progressBlock: ((Float) -> Void)? = nil,
        completeBlock: ((Error?) -> Void)? = nil
    ) -> LWFileDownloadTask {

        let manager = LWFileDownloadManager.shared

        let completeBlockWrapper: ((Error?) -> Void) = { error in
            manager.group.async(queue: manager.serialQueue) {
                completeBlock?(error)
            }
        }

        let task = LWFileDownloadTask(urlString: urlString, fileName: fileName)
        task.updateRequestBlock = requestBlock
        task.updateProgressBlock = progressBlock
        task.completeBlock = completeBlockWrapper

        return task
    }

    // MARK: - Helper Methods

    public var fileDirectoryPath: String {
        if let path = _fileDirectoryPath {
            return path
        }

        let documentsDirectory = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0]
        let path = (documentsDirectory as NSString).appendingPathComponent(directoryName)
        _fileDirectoryPath = path

        LWFileDownloadManager.createDirectory(ifNotExistsAtPath: path)

        return path
    }

    /// Create directory if it doesn't exist
    @objc @discardableResult
    public static func createDirectory(ifNotExistsAtPath path: String) -> Bool {
        let fileManager = FileManager.default

        if !fileManager.fileExists(atPath: path) {
            do {
                try fileManager.createDirectory(atPath: path, withIntermediateDirectories: true, attributes: nil)
                LWDLLog("Create directory Success!")
                return true
            } catch {
                LWDLLog("Error! %@", error.localizedDescription)
                return false
            }
        }
        return true
    }

    /// Get document directory path for a given path
    @objc public static func documentDirectoryPath(_ path: String) -> String {
        let documentsDirectory = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0]
        let directoryPath = (documentsDirectory as NSString).appendingPathComponent(path)

        createDirectory(ifNotExistsAtPath: directoryPath)
        return directoryPath
    }

    /// Write data to file path
    @objc @discardableResult
    public static func write(data: Data, toFilePath filePath: String) -> Bool {
        do {
            try data.write(to: URL(fileURLWithPath: filePath), options: .atomic)
            LWDLLog("=======file writeToFile: YES")
            return true
        } catch {
            LWDLLog("=======file writeToFile: NO - %@", error.localizedDescription)
            return false
        }
    }

    /// Check if file exists with fileName
    @objc public static func fileExists(withFileName fileName: String) -> Bool {
        let filePath = (LWFileDownloadManager.shared.fileDirectoryPath as NSString).appendingPathComponent(fileName)
        return FileManager.default.fileExists(atPath: filePath)
    }

    /// Get file path for fileName
    @objc public static func filePath(withFileName fileName: String) -> String {
        return (LWFileDownloadManager.shared.fileDirectoryPath as NSString).appendingPathComponent(fileName)
    }

    /// Remove file at path
    @objc @discardableResult
    public static func removeFile(atPath filePath: String) -> Bool {
        do {
            try FileManager.default.removeItem(atPath: filePath)
            return true
        } catch {
            LWDLLog("Error! %@", error.localizedDescription)
            return false
        }
    }
}
